Bitget: Add exchange support

cranktakular opened the pull request:
# PR Description

Implementing Bitget's V2 API, specifically the Common, Spot, Future, Margin, Earn, and Inst Loan sections.

In addition, does some smaller fixes across the codebase:
 - Exports more errors.
 - Lets types.Number support "null" values, interpreting them as 0.
 - Removes GetPositionSummary, as there's already GetFuturesPositionSummary, which fulfills the same functionality.

## Type of change

Please delete options that are not relevant and add an `x` in `[]` as item is complete.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## How has this been tested

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration and also consider improving test coverage whilst working on a certain feature or package.

- [x] go test ./... -race
- [x] golangci-lint run
- [x] Test X

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally and on Github Actions with my changes

cranktakular committed:
Initial bitget setup

cranktakular committed:
Bitget progress

cranktakular committed:
Bitget progress

cranktakular committed:
Bitget progress

cranktakular committed:
Bitget day 1 done

cranktakular committed:
Bitget progress

cranktakular committed:
Bitget progress

cranktakular committed:
Bitget progress and refinement

cranktakular committed:
Bitget mid-day progress

cranktakular committed:
More bitget progress

cranktakular committed:
Merge branch 'master' into bitget_api_creation

cranktakular committed:
Yet more Bitget progress

cranktakular committed:
Bitget progress and number expansion

cranktakular committed:
Mid-day bitget progress

cranktakular committed:
Another day, another Bitget

cranktakular committed:
Daily Bitget Progress

cranktakular committed:
Bitget progression

cranktakular committed:
Continual Bitget progress

cranktakular committed:
Bitget blitzing

cranktakular committed:
Bitget progress, REST almost done

cranktakular committed:
Bitget progress

cranktakular committed:
Bitget progress

cranktakular committed:
Little bit of bitget progress

cranktakular committed:
New endpoints & generic experiments

cranktakular committed:
Small fixes

cranktakular committed:
Bitget progress

cranktakular committed:
Merge branch 'master' into bitget_api_creation

cranktakular committed:
One xeno's paradox of progress

cranktakular committed:
Bitget building

cranktakular committed:
A bitget commit

cranktakular committed:
Bitget API becomes appended

cranktakular committed:
Mid-day commit

cranktakular committed:
Another bitget commit done

cranktakular committed:
Another Bitget commit deployed expertly

cranktakular committed:
Alas Bitget continues

cranktakular committed:
Bitget wrapper wrapup

cranktakular committed:
Bitget websocket beginnings

cranktakular committed:
Bitget websocket

cranktakular committed:
Bitget progression

cranktakular committed:
Websocket work & pre-emptive nit fixes

cranktakular committed:
Fixes, cleanup, and websocket progress

cranktakular committed:
Bitget interim commit

cranktakular committed:
Main websocket programming done

cranktakular committed:
Bitget is Sisyphean

cranktakular committed:
Bitget work

cranktakular committed:
Error name fixes

cranktakular committed:
Testing improvements

cranktakular committed:
More minor fixes

cranktakular committed:
Many minor fixes

cranktakular committed:
Approaching test perfection

cranktakular committed:
Big merge, no fixes yet

cranktakular committed:
A few fixes

cranktakular committed:
Post-merge improvements

cranktakular committed:
Merge branch 'master' into bitget_api_creation

cranktakular committed:
Improvements

cranktakular committed:
Frustrations

cranktakular committed:
Rate issue fixed

cranktakular committed:
Merge branch 'master' into bitget_api_creation

cranktakular committed:
Further improvements

cranktakular committed:
Merge branch 'master' into bitget_api_creation

cranktakular committed:
Bitget Sanity

cranktakular committed:
Bitget progresses swimmingly

cranktakular committed:
Bitget sanity improvements

cranktakular committed:
Bitget begets more Bitget

cranktakular committed:
Bitget sanity restored

cranktakular committed:
Wrapper improvements

cranktakular committed:
A few fixes

cranktakular committed:
Merge branch 'master' into bitget_api_creation

cranktakular committed:
Bitget done?!?

cranktakular committed:
Subscribing gooder

cranktakular committed:
Lint nits

cranktakular committed:
Test improvements

cranktakular committed:
Merge branch 'master' into bitget_api_creation

cranktakular committed:
Bitget progress

cranktakular committed:
Bitget updates

cranktakular committed:
Bitget doc check

cranktakular committed:
Merge branch 'master' into bitget_api_creation

cranktakular committed:
Lint nits

cranktakular committed:
Last-minute fixes

cranktakular committed:
Autofixes

cranktakular committed:
Further fixes

cranktakular committed:
Can't spell lint without nit

cranktakular committed:
Concat time

cranktakular committed:
Yet more fixes

cranktakular committed:
Merge branch 'master' into bitget_api_creation

cranktakular committed:
Final(?) Fixes

gloriousCode reviewed (changes requested):
Thank you for this PR! I look forward to going over it all.

cranktakular commented on exchanges/bitget/bitget_test.go:
```diff
@@ -0,0 +1,4009 @@
+	}
+}
+
+func getCrossOrdIDHelper(t *testing.T, skip bool) *OrderIDStruct {
+	t.Helper()
+	resp, err := bi.GetCrossOpenOrders(context.Background(), testPair, "", 0, 5, 1<<62, time.Now().Add(-time.Hour*24*85), time.Time{})
+	require.NoError(t, err)
+	if len(resp.OrderList) != 0 {
+		return &OrderIDStruct{
+			OrderID:       EmptyInt(resp.OrderList[0].OrderID),
+			ClientOrderID: resp.OrderList[0].ClientOrderID,
+		}
+	}
+	if skip {
+		t.Skip(skipInsufficientOrders)
+	}
+	return &OrderIDStruct{
+		OrderID:       0,
+		ClientOrderID: ordersNotFound,
+	}
+}
+
+func getIsoOrdIDHelper(t *testing.T, skip bool) *OrderIDStruct {
+	t.Helper()
+	resp, err := bi.GetIsolatedOpenOrders(context.Background(), testPair, "", 0, 5, 1<<62, time.Now().Add(-time.Hour*24*85), time.Time{})
+	require.NoError(t, err)
+	if len(resp.OrderList) != 0 {
+		return &OrderIDStruct{
+			OrderID:       EmptyInt(resp.OrderList[0].OrderID),
+			ClientOrderID: resp.OrderList[0].ClientOrderID,
+		}
+	}
+	if skip {
+		t.Skip(skipInsufficientOrders)
+	}
+	return &OrderIDStruct{
+		OrderID:       0,
+		ClientOrderID: ordersNotFound,
+	}
+}
```
Yeah these were all made 10 months ago, I'll start unwinding 'em.

cranktakular replied:
Done and pushed 🚀

qianniancn replied:
When can the merger be carried out

cranktakular committed:
Helper removal

cranktakular committed:
Merge branch 'master' into bitget_api_creation

cranktakular committed:
Bitget endpoint fixes

cranktakular committed:
Adding comment for a realisation

cranktakular committed:
Further fixes

cranktakular committed:
Merge branch 'master' into bitget_api_creation

cranktakular committed:
Post-merge fixes

cranktakular committed:
Config json lint fix

samuael commented on exchanges/bitget/bitget.go:
```diff
@@ -0,0 +1,5523 @@
+		params.Values.Set("limit", strconv.FormatInt(limit, 10))
+	}
+	params.Values.Set("orderId", strconv.FormatInt(orderID, 10))
+	path := bitgetSpot + bitgetTrade + "/" + bitgetFills
```
Rm this too

samuael reviewed (commented):
The implementation is very good. Good Job, man.
Just left some comments based on my observation.

I can review another round if/when you finish the changes

cranktakular commented on exchanges/bitget/bitget.go:
```diff
@@ -0,0 +1,5523 @@
+	if clientOrderID != "" {
+		req["clientOid"] = clientOrderID
+	}
+	path := bitgetSpot + bitgetTrade + bitgetPlacePlanOrder
```
This endpoint doesn't return any data; it signifies success/failure through their generic error code system.

https://www.bitget.com/api-doc/contract/account/Change-Margin

samuael replied:
I mean could you move the path to the call directly. That way you don't need to creaate the 'path' variable.

cranktakular committed:
Bitgets better

cranktakular committed:
Merge branch 'master' into bitget_api_creation

cranktakular committed:
Interface test causes lint issues and seems overall unnecessary

gbjk commented on exchanges/bitget/bitget.go:
```diff
@@ -0,0 +1,5523 @@
+	if multiCurrencyMode {
+		req["batchMode"] = "multiple"
+	}
+	path := bitgetSpot + bitgetTrade + bitgetBatchCancel
```
I actually think that stlyistically the current code is preferable.
Yes, it's a single use var, but it's a lot more legible.

You can make it a const, though, easily with a whole file regex replacement:
```
%s/^\tpath := bitget/\tconst path = bitget/
```
There's one time it needs to be a var, in `GetFuturesCandlestickData`
It might miss some examples but it's a good step.

Here's the change: gbjk@c65d5acd3c

You can run this, and then commit yourself:
```
curl -s https://github.com/gbjk/gocryptotrader/commit/c65d5acd3c.patch | git apply
```

Or You can yoink this commit:
```
curl -s https://github.com/gbjk/gocryptotrader/commit/c65d5acd3c.patch | git am
```

gbjk replied:
In today's episode of Fun with ~~Flags~~ Compilers, I tried to check if the var is going to be inlined:
`go build -gcflags="-m"` 
But I couldn't see any sign of it. Didn't sniff it too long.

cranktakular replied:
The compiler's smart enough; putting path as a const or a var results in the exact same assembly output. Checked using a particularly short and simple function (GetSupportedCurrencies).

samuael commented on exchanges/bitget/bitget.go:
```diff
@@ -0,0 +1,5105 @@
+		"passphrase":     passphrase,
+		"label":          label,
+		"ipList":         allowedIPList,
+		"permList":       permList,
```
Could you use a 
```
struct{
  SubAcctName string `json:"subAccountName"`
  .
  .
  .
}

```
and have tags like `omitempty` **JSON tag** to ignore zero values if not required in all such post methods?

cranktakular replied:
The bodyParams parameter of SendAuthenticatedHTTPRequest() isn't set up for that. I guess I _could_ replace that field with an interface; do you think that's better?

samuael commented on exchanges/bitget/bitget.go:
```diff
@@ -0,0 +1,5105 @@
+		"passphrase":    passphrase,
+		"label":         label,
+		"ipList":        whiteList,
+		"permList":      permList,
```
Same goes here and all other such payloads(arguments)

samuael commented on exchanges/bitget/bitget.go:
```diff
@@ -0,0 +1,5105 @@
+	if amount <= 0 {
+		return nil, order.ErrAmountBelowMin
+	}
+	req := map[string]any{
```
Can easily be solved by using a `struct` than a `map[string]any`

samuael commented on exchanges/bitget/bitget.go:
```diff
@@ -0,0 +1,5523 @@
+		return nil, errPairEmpty
+	}
+	if len(orders) == 0 {
+		return nil, errOrdersEmpty
```
You can add the further explanation about the error by wrapping the `common.Err...` with the fmt.Errorf("%w ...."). It's up to you to chose though

samuael commented on exchanges/bitget/bitget.go:
```diff
@@ -0,0 +1,5523 @@
+		return nil, errPairEmpty
+	}
+	if side == "" {
+		return nil, errSideEmpty
```
You decided to leave it lke this or didn't have time to make update?

cranktakular replied:
I implemented the Errorf stuff in the variable declaration for errSideEmpty (and other similar errors) at the top of the file, rather than in each instance of usage, to save on copy/pasting.

samuael commented on exchanges/bitget/bitget.go:
```diff
@@ -0,0 +1,5523 @@
+		return nil, errTriggerPriceEmpty
+	}
+	if orderType == "" {
+		return nil, errOrderTypeEmpty
```
?

samuael reviewed (commented):
Thank you @cranktakular  for the updates.
I have added few more reviews for now and will try to add when you finish this round of update.

Additionally, there is a single linter issue you may need to fix

cranktakular commented on exchanges/bitget/bitget.go:
```diff
@@ -0,0 +1,5105 @@
+// GetP2PMerchantList returns detailed information on merchants
+func (bi *Bitget) GetP2PMerchantList(ctx context.Context, online string, limit, pagination int64) (*P2PMerListResp, error) {
+	vals := url.Values{}
+	vals.Set("online", online)
```
For this and others, I only check for zero values if (at the time of developing the function) the exchange marks the field as required, breaks, or operates strangely with the zero value being sent.

cranktakular committed:
Unmarshaling improvements

cranktakular committed:
Merge and associated fixes

cranktakular committed:
Lint and test fixes

cranktakular committed:
Merge and post-merge fixes

cranktakular committed:
Assert nit

cranktakular committed:
Merge and post-merge fixes

cranktakular committed:
Merge branch 'master' into bitget_api_creation

cranktakular committed:
Incremental improvements

cranktakular committed:
Merge branch 'master' into bitget_api_creation

cranktakular committed:
Finalisation

cranktakular committed:
Removing unused test constants

cranktakular committed:
Merge branch 'master' into bitget_api_creation

cranktakular committed:
Post-merge fixes

thrasher- commented:
@codex please review

cranktakular committed:
Bitget API updates

cranktakular committed:
Bitget API corrections

cranktakular committed:
Bitget-focused adjustments to our engine

cranktakular committed:
Merge branch 'master' into bitget_api_creation

cranktakular committed:
Bitget adjustments in line with API updates

samuael commented on exchanges/bitget/bitget.go:
```diff
@@ -0,0 +1,5033 @@
+	errToIDEmpty                      = errors.New("toID cannot be empty")
+	errTransferTypeEmpty              = errors.New("transferType cannot be empty")
+	errAddressEmpty                   = errors.New("address cannot be empty")
+	errMarginCoinEmpty                = errors.New("marginCoin cannot be empty")
```
```suggestion
	errMarginCoinEmpty                = fmt.Errrof("%w: marginCoin cannot be empty", currency.ErrCurrencyCodeEmpty)
```

samuael commented on exchanges/bitget/bitget.go:
```diff
@@ -0,0 +1,5033 @@
+	errLeverageEmpty                  = errors.New("leverage cannot be empty")
+	errMarginModeEmpty                = fmt.Errorf("%w margin mode can not be empty", margin.ErrInvalidMarginType)
+	errPositionModeEmpty              = errors.New("positionMode cannot be empty")
+	errNewClientOrderIDEmpty          = errors.New("newClientOrderID cannot be empty")
```
```suggestion
	errNewClientOrderIDEmpty          = fmt.Errrof("%w: newClientOrderID cannot be empty", order.ErrClientOrderIDMustBeSet)
```

samuael commented on exchanges/bitget/bitget.go:
```diff
@@ -0,0 +1,5033 @@
+	if productType != "" {
+		params.Values.Set("productType", productType)
+	}
+	params.Values.Set("marginCoin", currency.String())
```
This is not checked for zero value.

samuael commented on exchanges/bitget/bitget.go:
```diff
@@ -0,0 +1,5033 @@
+	params.Values.Set("annType", annType)
+	params.Values.Set("language", "en_US")
+	params.Values.Set("idLessThan", strconv.FormatInt(pagination, 10))
+	if limit != 0 {
```
check if the `limit` is `> 0`, not `!= 0` the use `strconv.FormatInt()` instead.

samuael commented on exchanges/bitget/bitget.go:
```diff
@@ -0,0 +1,5033 @@
+// GetTime returns the server's time
+func (e *Exchange) GetTime(ctx context.Context) (*TimeResp, error) {
+	var resp TimeResp
+	return &resp, e.SendHTTPRequest(ctx, exchange.RestSpot, Rate20, bitgetPublic+bitgetTime, nil, &resp)
```
use *TimeResp and 
```suggestion
	return resp, e.SendHTTPRequest(ctx, exchange.RestSpot, Rate20, bitgetPublic+bitgetTime, nil, &resp)
```

samuael reviewed (commented):
I will do another round of review when these comments are actions.
Thanks

cranktakular committed:
Merge branch 'master' into bitget_api_creation

cranktakular committed:
Fixes post-merge issues / Corrects issues identified by review / Fixes discovered exchange-side errors

shazbert commented on exchanges/bitget/bitget_websocket.go:
```diff
@@ -0,0 +1,1212 @@
+
+// CandleDataHandler handles candle data, as functionality is shared between updates and snapshots
+func (e *Exchange) candleDataHandler(wsResponse *WsResponse) error {
+	var candles [][8]string
```
```suggestion
	var candles [][8]types.Number
```

This might work; I understand the int part but its a timestamp it shouldn't overflow.

shazbert commented on exchanges/bitget/bitget_websocket.go:
```diff
@@ -0,0 +1,1212 @@
+
+// TradeDataHandler handles trade data, as functionality is shared between updates and snapshots
+func (e *Exchange) tradeDataHandler(wsResponse *WsResponse) error {
+	var trades []WsTradeResponse
```
hug with Unmarshal across this file

shazbert commented on exchanges/bitget/bitget_websocket.go:
```diff
@@ -0,0 +1,1212 @@
+	if err != nil {
+		return err
+	}
+	err = json.Unmarshal(wsResponse.Data, &trades)
```
if short as much as you can across this PR

shazbert commented on exchanges/bitget/bitget_websocket.go:
```diff
@@ -0,0 +1,1212 @@
+// OrderbookDataHandler handles orderbook data, as functionality is shared between updates and snapshots
+func (e *Exchange) orderbookDataHandler(wsResponse *WsResponse) error {
+	var ob []WsOrderBookResponse
+	pair, err := pairFromStringHelper(wsResponse.Arg.InstrumentID)
```
suggestion: This looks like it could be a type with Unmarshaller but up to you

shazbert commented on exchanges/bitget/bitget_websocket.go:
```diff
@@ -0,0 +1,1212 @@
+			ValidateOrderbook:      e.ValidateOrderbook,
+			ChecksumStringRequired: true,
+		}
+		err = e.Websocket.Orderbook.LoadSnapshot(&ob)
```
can just return this and then drop the else

cranktakular replied:
The else is to distinguish snapshot and update so I can't drop it, but you're right that I can just return that.

shazbert commented on exchanges/bitget/bitget_websocket.go:
```diff
@@ -0,0 +1,1212 @@
+			ExpectedChecksum: uint32(ob[0].Checksum), //nolint:gosec // The exchange sends it as ints expecting overflows to be handled as Go does by default
+		}
+		// Sometimes the exchange returns updates with no new asks or bids, just a checksum and timestamp
+		if len(update.Bids) != 0 || len(update.Asks) != 0 {
```
Can drop this and Use field AllowEmpty as those will bump the LastUpdateID and it just lets you know that no action happened on the books between the update frequency.

shazbert commented on exchanges/bitget/bitget_websocket.go:
```diff
@@ -0,0 +1,1212 @@
+}
+
+func (e *Exchange) fillDataHandler(wsResponse *WsResponse, respRaw []byte) error {
+	respAsset := itemDecoder(wsResponse.Arg.InstrumentType)
```
suggestion: Seems like this could be a type with unmarshaller but up to you

shazbert commented on exchanges/bitget/bitget_websocket.go:
```diff
@@ -0,0 +1,1212 @@
+		}
+		e.Websocket.DataHandler <- resp
+	default:
+		e.Websocket.DataHandler <- websocket.UnhandledMessageWarning{Message: e.Name + websocket.UnhandledMessage + string(respRaw)}
```
This looks like an unhandled asset error and could drop the respRaw from param

shazbert commented on exchanges/bitget/bitget_websocket.go:
```diff
@@ -0,0 +1,1212 @@
+			LastUpdated:  indexPrice[i].Timestamp.Time(),
+		}
+	}
+	resp = resp[:cur]
```
This does look a bit weird im going to have to test this one out in the future. 😆

shazbert commented on exchanges/bitget/bitget_websocket.go:
```diff
@@ -0,0 +1,1212 @@
+// ReqBuilder builds a request in the manner the exchange expects
+func (e *Exchange) reqBuilder(req *WsRequest, sub *subscription.Subscription) {
+	for i := range sub.Pairs {
+		form, _ := e.GetPairFormat(asset.Spot, true)
```
handle error

shazbert commented on exchanges/bitget/bitget_websocket.go:
```diff
@@ -0,0 +1,1212 @@
+	for i := range unauthReq {
+		if len(unauthReq[i].Arguments) != 0 {
+			wg.Add(1)
+			go func(req WsRequest) {
```
can use wg.Go and drop the Add and defer done.

shazbert commented on exchanges/bitget/bitget_websocket.go:
```diff
@@ -0,0 +1,1212 @@
+	return template.New("master.tmpl").Funcs(template.FuncMap{"channelName": channelName}).Parse(subTplText)
+}
+
+func channelName(s *subscription.Subscription) string {
```
can return an error without ill effect and not panic inside the function.

shazbert commented on exchanges/bitget/bitget_websocket.go:
```diff
@@ -0,0 +1,1212 @@
+}
+
+// GetSubscriptionTemplate returns a subscription channel template
+func (e *Exchange) GetSubscriptionTemplate(_ *subscription.Subscription) (*template.Template, error) {
```
```suggestion
func (e *Exchange) GetSubscriptionTemplate(*subscription.Subscription) (*template.Template, error) {
```

shazbert commented on exchanges/bitget/bitget_wrapper.go:
```diff
@@ -0,0 +1,2991 @@
+	return nil
+}
+
+// FetchTicker returns the ticker for a currency pair
+func (e *Exchange) FetchTicker(ctx context.Context, p currency.Pair, assetType asset.Item) (*ticker.Price, error) {
+	tickerNew, err := ticker.GetTicker(e.Name, p, assetType)
+	if err != nil {
+		return e.UpdateTicker(ctx, p, assetType)
+	}
+	return tickerNew, nil
+}
+
+// FetchOrderbook returns orderbook base on the currency pair
+func (e *Exchange) FetchOrderbook(ctx context.Context, pair currency.Pair, assetType asset.Item) (*orderbook.Book, error) {
+	ob, err := orderbook.Get(e.Name, pair, assetType)
+	if err != nil {
+		return e.UpdateOrderbook(ctx, pair, assetType)
+	}
+	return ob, nil
+}
```
These can be removed

shazbert commented on exchanges/bitget/bitget_wrapper.go:
```diff
@@ -0,0 +1,2991 @@
+}
+
+// ClientIDGenerator is a helper function that generates a unique client ID
+func clientIDGenerator() string {
```
use base method `MessageID` which has a uuid

shazbert commented on exchanges/bitget/bitget_wrapper.go:
```diff
@@ -0,0 +1,2991 @@
+	case kline.OneMonth:
+		return "1Mutc"
+	}
+	return errIntervalNotSupported
```
can wrap these errors with incoming value and below

shazbert commented on exchanges/bitget/ratelimit.go:
```diff
@@ -0,0 +1,45 @@
+)
+
+// GetRateLimits returns the rate limits for Bitget
+func GetRateLimits() request.RateLimitDefinitions {
```
Needs to be a var for multiple allocations of Exchange so they can share rate limits.

shazbert reviewed (changes requested):
Very fine work. Some quick nits to get started.

cranktakular commented on cmd/exchange_wrapper_issues/main.go:
```diff
@@ -912,24 +912,6 @@ func testWrappers(e exchange.IBotExchange, base *exchange.Base, cfg *Config) []E
 			Response:   marginRateHistoryResponse,
 		})
 
-		positionSummaryRequest := &futures.PositionSummaryRequest{
```
As explained at the top of the PR:
`Removes GetPositionSummary, as there's already GetFuturesPositionSummary, which fulfills the same functionality.`

shazbert commented on exchanges/bitget/bitget.go:
```diff
@@ -0,0 +1,5329 @@
+	// Public endpoints
+	bitgetPublic                   = "public/"
+	bitgetAnnouncements            = "annoucements" // Misspelling of announcements
+	bitgetTime                     = "time"
```
Yeah that's fine keep as is. 👍

cranktakular commented on exchanges/bitget/bitget.go:
```diff
@@ -0,0 +1,5329 @@
+}
+
+// PlaceSpotOrder places a spot order on the exchange
+func (e *Exchange) PlaceSpotOrder(ctx context.Context, pair currency.Pair, side, orderType, strategy, clientOrderID, stpMode string, price, amount, triggerPrice, presetTPPrice, executeTPPrice, presetSLPrice, executeSLPrice float64, isCopyTradeLeader bool, acceptableDelay time.Duration) (*OrderIDStruct, error) {
```
What do you mean by "and below"? Just that you'd also ask this of other functions with big anonymous structs?

cranktakular committed:
Applies changes requested by shazbert

cranktakular commented on exchanges/bitget/bitget_test.go:
```diff
@@ -0,0 +1,4082 @@
+	t.Parallel()
+	_, err := e.GetFuturesActiveVolume(t.Context(), currency.Pair{}, "")
+	assert.ErrorIs(t, err, currency.ErrCurrencyPairEmpty)
+	te := new(Exchange) // The endpoint intermittently returns "The data fetched by BTCUSDT is empty", while otherwise accepting that valid request. Mocking to avoid that flakiness.
```
Why do we do that?

shazbert replied:
stops copy/pasting issues and is consistent to what we are currently doing across other exchanges

cranktakular commented on exchanges/bitget/bitget_test.go:
```diff
@@ -0,0 +1,4082 @@
+
+func TestPlaceTriggerFuturesOrder(t *testing.T) {
+	t.Parallel()
+	_, err := e.PlaceTriggerFuturesOrder(t.Context(), "", "", "", "", "", "", "", "", "", "", "", currency.Pair{}, currency.Code{}, 0, 0, 0, 0, 0, 0, 0, 0, false)
```
That seems like it would've had to have been quite contrived to set up with the way it was previously. With the new change to a single point parameter that does seem like a more feasible change.

cranktakular committed:
Applies another batch of changes requested by shazbert

samuael reviewed (commented):
This is a cool PR. I have made a quick review. Will get back to it another round

samuael reviewed (commented):
Just added few comments on top of what I did before.

samuael reviewed (changes requested):
Not all of my last review are applied but I have added more

cranktakular committed:
Implements some requested changes

cranktakular commented on exchanges/orderbook/orderbook_types.go:
```diff
@@ -21,13 +21,13 @@ const (
 // Public errors
 var (
 	ErrOrderbookNotFound = errors.New("cannot find orderbook(s)")
+	ErrAssetTypeNotSet   = errors.New("orderbook asset type not set")
```
I think it makes some sense; the orderbook's asset type isn't set, seems fine for the orderbook package.

But I wouldn't be too opposed to moving it.

cranktakular committed:
Implements more requested changes

cranktakular committed:
Implements even more requested changes

samuael commented on exchanges/bitget/bitget.go:
```diff
@@ -0,0 +1,4840 @@
+// QueryAnnouncements returns announcements from the exchange, filtered by type and time
+func (e *Exchange) QueryAnnouncements(ctx context.Context, annType string, startTime, endTime time.Time, pagination uint64, limit uint8) ([]AnnouncementResp, error) {
+	var params Params
+	params.Values = make(url.Values)
```
move this to below the error check.

samuael commented on exchanges/bitget/bitget.go:
```diff
@@ -0,0 +1,4840 @@
+	}
+	vals := url.Values{}
+	vals.Set("businessType", businessType)
+	var resp []AllTradeRatesResp
```
can use []* for all response type declarations from now on; for consistency.
Instead of iterating over a slice and use the index to access elements, we can iterate over the slice but use the pointer and reduce the memory overhead.

samuael commented on exchanges/bitget/bitget.go:
```diff
@@ -0,0 +1,4840 @@
+	}
+	vals := url.Values{}
+	vals.Set("symbol", pair.String())
+	vals.Set("period", period)
```
`period` is not checked for zero value

samuael commented on exchanges/bitget/bitget.go:
```diff
@@ -0,0 +1,4840 @@
+	if period != "" {
+		vals.Set("period", period)
+	}
+	vals.Set("coin", cur.String())
```
`cur` is not checked if it is zero value.

samuael reviewed (commented):
**Shalooom** 🕊️  brother.
Just added another round of review on endpoint functions.

shazbert reviewed (changes requested):
Will start to integrate this as priority and will help out with any changes

Commit:
Increases MaxWebsocketSubscriptionsPerConnection from 240 to 1000 as per docs

Commit:
Merge branch 'master' into bitget_api_creation

Commit:
bitget: after merge update for accounts and futures

Commit:
linter: fix some issues

Commit:
bitget: fix tests and refactor

Commit:
bitget: use internal orderbook type to reduce code footprint

Commit:
bitget: migrate to websocket multi-connection

Commit:
bitget: link up config subs with subscription expansion, enable all default subs

shazbert commented on exchanges/bitget/bitget.go:
```diff
@@ -0,0 +1,4839 @@
+
+// QueryAnnouncements returns announcements from the exchange, filtered by type and time
+func (e *Exchange) QueryAnnouncements(ctx context.Context, annType string, startTime, endTime time.Time, pagination uint64, limit uint8) ([]AnnouncementResp, error) {
+	var params Params
```
suggestion: Params type can have url.Values unexported and then just have methods to the underlying which calls make when there is a nil reference, stops you having to do `params.Values = make(url.Values)`

shazbert reviewed (changes requested):
Have updated accounts which needs more testing, migrated to multi-connection and added some fixes to subs. On login for the private conn it waits until it has a confirmation message. 

Some things that need to be addressed:

* futures asset needs to be broken up to USDT margined, USDC margined and Coin Margined futures. 
* margin and cross margin subs are semi broken
* linter/misc/spell needs fixing
* bitget config needs to be added to config_example.json


Will keep pushing things as I came across them. 🚀

Commit:
biget: fix fetch tradable pairs for futures, and update execution handling, plus todos

cranktakular commented on exchanges/bitget/bitget.go:
```diff
@@ -0,0 +1,4909 @@
+		vals.Set("period", period)
+	}
+	path := bitgetMargin + bitgetMarket + bitgetIsolatedBorrowRate
+	var resp []BorrowRatioResp
```
What do you mean? It's slightly easier to loop over slices when they aren't of pointers since then you don't need to dereference it afterwards.
```
sli := []int64{1, 2, 3, 4}
	for _, b := range sli {
		fmt.Printf("%v\n", b)
	}
```
Compared to
```
i1, i2, i3, i4 := int64(1), int64(2), int64(3), int64(4)
	sli := []*int64{&i1, &i2, &i3, &i4}
	for _, b := range sli {
		fmt.Printf("%v\n", *b)
	}
```

cranktakular replied:
You need to dereference it in order to do most things with it. If you don't, then you just have an arbitrary pointer. That means nothing until you dereference it to get the underlying data. (Although, you can send it off to another function without dereferencing; that function will just need to dereference it in order to make use of it)

cranktakular commented on exchanges/bitget/bitget.go:
```diff
@@ -0,0 +1,4909 @@
+		return nil, errSubaccountEmpty
+	}
+	path := bitgetUser + bitgetCreate + bitgetVirtualSubaccount
+	req := struct {
```
They were maps; you [told me to replace them with structs](https://github.com/thrasher-corp/gocryptotrader/pull/1844#discussion_r2080459765) back in May!

cranktakular committed:
Implements requested changes, fixes bugs

Commit:
feat(bitget): implement websocket order placement and cancellation

- Added websocket support for placing spot and futures orders with validation checks.
- Implemented cancellation of orders via websocket for both spot and futures.
- Created request and response types for websocket interactions.
- Added tests for websocket order submission and cancellation to ensure proper functionality.
- Updated order types to include settlement currency for derivatives.

Commit:
Merge branch 'master' into bitget_api_creation

cranktakular committed:
Merge branch 'master' into bitget_api_creation

cranktakular committed:
Fixes some issues, implements harnesses for exploring others

cranktakular committed:
Fixes some test failures, removes some debugging code

cranktakular committed:
Merge branch 'master' into bitget_api_creation

cranktakular committed:
Fixes some test and lint failures, restores some clarifying comments, conforms implementations with updates to endpoints

cranktakular committed:
Merge branch 'master' into bitget_api_creation

cranktakular committed:
Tweaks websocket management to remove bugs, adds helpers for context debugging, implements a new endpoint, fixes some other bugs and inconsistencies

cranktakular committed:
Merge branch 'master' into bitget_api_creation

cranktakular committed:
Improves coverage, implements a more appropriate sandbox field, adopts more exchange functionality

Commit:
Merge branch 'master' into bitget_api_creation

samuael commented on exchanges/bitget/bitget_websocket_test.go:
```diff
@@ -0,0 +1,457 @@
+
+func TestWsConnect(t *testing.T) {
+	// exch := &Exchange{}
+	// exch.Websocket = sharedtestvalues.NewTestWebsocket()
```
Please remove

samuael commented on exchanges/bitget/bitget_websocket_test.go:
```diff
@@ -0,0 +1,457 @@
+}
+
+func TestWsAuth(t *testing.T) {
+	// e.Websocket.SetCanUseAuthenticatedEndpoints(false)
```
Here too

samuael commented on exchanges/bitget/bitget_websocket_test.go:
```diff
@@ -0,0 +1,457 @@
+}
+
+// func TestWsReadData(t *testing.T) {
+// 	mock := func(tb testing.TB, msg []byte, w *gws.Conn) error {
```
And here

samuael commented on exchanges/bitget/bitget_websocket_test.go:
```diff
@@ -0,0 +1,457 @@
+
+func TestTickerDataHandler(t *testing.T) {
+	t.Parallel()
+	mockJSON := []byte(`{"event":"snapshot","arg":{"channel":"ticker","instType":"SPOT"},"data":[[]]}`)
```
Could you group these JSONs into a single list and run the test by looping ober these or use t.Run()

samuael commented on exchanges/bitget/bitget_websocket_test.go:
```diff
@@ -0,0 +1,457 @@
+	mockJSON = []byte(`{"event":"snapshot","arg":{"channel":"candle1D","instId":"BTCUSD"},"data":[["1","a","3","4","5","6","",""]]}`)
+	err = e.wsHandleData(t.Context(), fixtureConnection{}, mockJSON)
+	assert.ErrorIs(t, err, strconv.ErrSyntax)
+	mockJSON = []byte(`{"event":"snapshot","arg":{"channel":"candle1D","instId":"BTCUSD"},"data":[["1","2","a","4","5","6","",""]]}`)
```
JSON as a list and iterate here too

samuael commented on exchanges/bitget/bitget_websocket_test.go:
```diff
@@ -0,0 +1,457 @@
+
+func TestAccountSnapshotDataHandler(t *testing.T) {
+	ctx := accounts.DeployCredentialsToContext(t.Context(), &accounts.Credentials{
+		Key:      "TestAccountSnapshotDataHandler",
```
Could you change the strjng contents tobsomethjng like "api-key-ges here". It makes it meaningful than just using the function name inplace of the values

samuael commented on exchanges/bitget/bitget_websocket_test.go:
```diff
@@ -0,0 +1,457 @@
+	err := e.wsHandleData(t.Context(), fixtureConnection{}, mockJSON)
+	assert.ErrorContains(t, err, errUnmarshalArray)
+	ctx := accounts.DeployCredentialsToContext(t.Context(), &accounts.Credentials{
+		Key:      "TestCrossAccountDataHandler",
```
Replace the string content with meaningful string to correctly explain the credentials

samuael commented on exchanges/bitget/bitget_websocket_test.go:
```diff
@@ -0,0 +1,457 @@
+	ctx := accounts.DeployCredentialsToContext(t.Context(), &accounts.Credentials{
+		Key:      "TestIsolatedAccountDataHandler",
+		Secret:   "TestIsolatedAccountDataHandler",
+		ClientID: "TestIsolatedAccountDataHandler",
```
Here too

samuael commented on exchanges/bitget/bitget_wrapper_test.go:
```diff
@@ -0,0 +1,73 @@
+	err = e.UpdateTickers(t.Context(), asset.USDTMarginedFutures)
+	assert.NoError(t, err)
+	err = e.UpdateTickers(t.Context(), asset.USDCMarginedFutures)
+	assert.NoError(t, err)
```
Use list and iterate over the values

samuael commented on exchanges/bitget/bitget_wrapper_test.go:
```diff
@@ -0,0 +1,73 @@
+	err = e.UpdateOrderExecutionLimits(t.Context(), asset.Spot)
+	assert.NoError(t, err)
+	err = e.UpdateOrderExecutionLimits(t.Context(), asset.USDCMarginedFutures)
+	assert.NoError(t, err)
```
Use []asset.Item and iterate over the values than repeating the method call declaration

samuael commented on exchanges/bitget/ratelimit.go:
```diff
@@ -0,0 +1,42 @@
+	rate10
+	rate5
+	rate3
+	rate2
```
This EndpointLimit values could have a better name

cranktakular replied:
Such as? There's no consistent kind of endpoint these values apply to, so I can only imagine naming them based on the limit.

samuael commented on exchanges/bitget/ws_requests.go:
```diff
@@ -0,0 +1,242 @@
+		return nil, err
+	}
+	if len(result) != 1 {
+		return nil, common.ErrNoResponse
```
Invalid or NoResponse??

samuael commented on exchanges/bitget/ws_requests.go:
```diff
@@ -0,0 +1,242 @@
+		price = strconv.FormatFloat(req.Price, 'f', -1, 64)
+	}
+
+	var result []WebsocketSpotPlaceOrderResponse
```
You can declare this as `[]*WebsocketSpotPlaceOrderResponse` and since it is efficient

cranktakular replied:
I don't think that's more efficient. And we typically do that when we're returning or iterating over a slice. Since that isn't done here, I don't think that change is positive.

samuael commented on exchanges/bitget/ws_requests.go:
```diff
@@ -0,0 +1,242 @@
+		ID:             strconv.FormatInt(e.GetBase().MessageSequence(), 10),
+		InstrumentType: "SPOT",
+		InstrumentID:   req.Pair.String(),
+		Channel:        "place-order",
```
Declare as constant

samuael commented on exchanges/bitget/ws_requests.go:
```diff
@@ -0,0 +1,242 @@
+	}
+
+	var price string
+	if req.Price > 0 {
```
You can add a check making price required for limit order types

cranktakular replied:
That was already done, a few lines above.

samuael commented on exchanges/bitget/ws_requests.go:
```diff
@@ -0,0 +1,242 @@
+		price = strconv.FormatFloat(req.Price, 'f', -1, 64)
+	}
+
+	var result []WebsocketSpotPlaceOrderResponse
```
Slice of pointer here

samuael commented on exchanges/bitget/ws_requests.go:
```diff
@@ -0,0 +1,242 @@
+	if req.MarginMode == "" {
+		return nil, errMarginModeUnset
+	}
+	if req.ReduceOnly == "" && req.TradeSide == "" {
```
Wrap this and the following error checks under

`If req.ReduceOnly == "" {
    // Here ...
}`

cranktakular replied:
No can do, they're checking different things.

Only the first one checks if ReduceOnly is empty, the second one checks if it's _not_ empty.

samuael commented on exchanges/bitget/ws_requests.go:
```diff
@@ -0,0 +1,242 @@
+		stopLoss = strconv.FormatFloat(req.PresetStopLossPrice, 'f', -1, 64)
+	}
+
+	var result []WebsocketFuturesPlaceOrderResponse
```
Same here

samuael commented on exchanges/bitget/ws_requests_test.go:
```diff
@@ -0,0 +1,257 @@
+		Price:       100,
+	})
+	require.NoError(t, err)
+	require.NotEmpty(t, got)
```
```suggestion
	assert.NotEmpty(t, got)
```

samuael commented on exchanges/bitget/ws_requests_types.go:
```diff
@@ -0,0 +1,130 @@
+type WebsocketFuturesOrderParams struct {
+	OrderType              string `json:"orderType"`
+	Side                   string `json:"side"`
+	Size                   string `json:"size"`
```
You could use float64 and use a ,string tag for when sending the JSON. For all occurences of of size, price and the likes of JSON payload

samuael commented on exchanges/bitget/ws_wrapper.go:
```diff
@@ -0,0 +1,135 @@
+			return nil, fmt.Errorf("%w: %q", margin.ErrInvalidMarginType, submit.MarginType)
+		}
+
+		// TODO:
```
There has to be an authenticated endpoint to check this account's information.

cranktakular replied:
Could you please clarify what you mean by this?

samuael commented on exchanges/bitget/ws_wrapper.go:
```diff
@@ -0,0 +1,135 @@
+		if err != nil {
+			return nil, err
+		}
+		sr, err := submit.DeriveSubmitResponse(resp.ID)
```
```suggestion
		return submit.DeriveSubmitResponse(resp.ID)
```

samuael commented on exchanges/bitget/ws_wrapper_test.go:
```diff
@@ -0,0 +1,124 @@
+		MarginType:         margin.Isolated,
+	})
+	require.NoError(t, err)
+	require.NotEmpty(t, got)
```
```suggestion
	assert.NotEmpty(t, got)
```
For all occurance of such final line assertion statements

samuael commented on exchanges/orderbook/orderbook.go:
```diff
@@ -87,7 +87,7 @@ func (s *store) DeployDepth(exchange string, p currency.Pair, a asset.Item) (*De
 	if !a.IsValid() {
-		return nil, errAssetTypeNotSet
+		return nil, ErrAssetTypeNotSet
 	}
```
There has to be such error declaration in the common package

cranktakular replied:
Not as far as I can see.

samuael reviewed (changes requested):
Nice work, sir. 
I have left comments on some code sanity and efficiency related issues.

cranktakular committed:
Merge branch 'master' into bitget_api_creation

cranktakular committed:
Addresses various issues pointed out by reviewers

cranktakular committed:
Corrects mistakes in line with review comments

cranktakular committed:
Merge branch 'master' into bitget_api_creation

gloriousCode commented on exchanges/bitget/ws_requests.go:
```diff
@@ -0,0 +1,224 @@
+	if req.MarginMode == "" {
+		return nil, errMarginModeUnset
+	}
+	if req.ReduceOnly == "" && req.TradeSide == "" {
```
This validation is stricter than Bitget's websocket futures order API. The official Place Order Channel docs list `tradeSide` as optional and only required in hedge mode, while `reduceOnly` is optional, defaults to `NO`, and applies only in one-way mode: https://www.bitget.com/api-doc/contract/websocket/private/Place-Order-Channel. Requiring exactly one of these rejects valid one-way open orders where both values are omitted, and it also diverges from the REST wrapper below. Please allow both to be empty and only reject combinations Bitget actually rejects.

gloriousCode commented on exchanges/bitget/bitget_types.go:
```diff
@@ -0,0 +1,3180 @@
+	PosLimit                  types.Number  `json:"posLimit"`
+	MaintainTime              types.Time    `json:"maintainTime"`
+	OpenTime                  types.Time    `json:"openTime"`
+	MaxMarketOrderQuantity    types.Number  `json:"maxMarketOrderQuantity"`
```
This does not match the futures contract config response. The docs and live `/api/v2/mix/market/contracts` response use `maxMarketOrderQty` and `maxOrderQty`, not `maxMarketOrderQuantity`, so `MaxMarketOrderQuantity` currently decodes as zero and the limit-order max is unavailable. Please change the tag to `json:"maxMarketOrderQty"`, add a `MaxOrderQuantity`/similar field for `maxOrderQty`, and cover both fields with a response fixture test.

cranktakular replied:
Not exactly sure what you mean by "response fixture test"; let me know if this change suffices.

gloriousCode commented on exchanges/bitget/bitget.go:
```diff
@@ -0,0 +1,4814 @@
+
+// PlaceFuturesOrder places a futures order on the exchange
+func (e *Exchange) PlaceFuturesOrder(ctx context.Context, p *PlaceSingleFuturesOrderParams, isCopyTradeLeader bool) (*OrderIDStruct, error) {
+	if p.Pair.IsEmpty() {
```
This dereferences `p` before checking for nil. Nearby methods such as `PlaceSpotOrder` return `common.ErrNilPointer` for nil request structs, and wrapper tests usually exercise nil request paths. Please add the same nil guard here so caller mistakes return an error instead of panicking.

gloriousCode commented on exchanges/bitget/bitget_websocket_test.go:
```diff
@@ -0,0 +1,382 @@
+}
+
+func TestWsAuth(t *testing.T) {
+	e.Websocket.SetCanUseAuthenticatedEndpoints(false)
```
These websocket handler tests would be easier to extend if they followed the newer Gateio-style local test setup instead of mutating the package-level `e`. For example, creating a fresh exchange with `sharedtestvalues.NewTestWebsocket()`/local `DataHandler` per test would let each handler assert the parsed payload sent to `DataHandler`, rather than mostly checking global-state `wsHandleData` error paths. That would make the Bitget websocket tests line up better with Gateio patterns and give us dedicated parsing coverage for each channel.

cranktakular replied:
Sure, but is it worthwhile swapping over to that at this point?

gloriousCode reviewed (changes requested):
Very basic review for you.

Local checks on PR head:
- `go test -race ./exchanges/bitget -count=1` passed.
- `make lint` failed with `exchanges/bitget/ws_requests.go:188:67: (*Exchange).sendWebsocketTradeRequest - epl always receives request.Unset (0) (unparam)`.
- `make misc_checks` failed because `exchanges/bitget/bitget_websocket_test.go` uses formatted messages with non-`f` testify helpers at lines 90, 113, 137, 171, 192, 213, 235, 253, 322, and 354.

cranktakular committed:
Merge branch 'master' into bitget_api_creation

cranktakular committed:
Fixes issues picked up by gloriousCode and automated tests
